services/api/service.go

```diff
@@ -825,11 +837,32 @@ func (api *RelayAPI) updateProposerDuties(headSlot uint64) {
 	api.log.Infof("proposer duties updated: %s", strings.Join(_duties, ", "))
 }
 
-func (api *RelayAPI) prepareBuildersForSlot(headSlot uint64) {
-	// Wait until there are no optimistic blocks being processed. Then we can
-	// safely update the slot.
+func (api *RelayAPI) prepareBuildersForSlot(headSlot, prevHeadSlot uint64) {
+	// First wait for this process to finish processing optimistic blocks
 	api.optimisticBlocksWG.Wait()
+
+	// Now we release our lock and wait for all other builder processes to wrap up
+	err := api.redis.EndProcessingSlot(context.Background())
+	if err != nil {
+		api.log.WithError(err).Error("failed to unlock redis optimistic processing slot")
+	}
+	err = api.redis.WaitForSlotComplete(context.Background(), prevHeadSlot+1)
+	if err != nil {
+		api.log.WithError(err).Error("failed to get redis optimistic processing slot")
+	}
+
+	// Prevent race with StopServer, make sure we don't lock up redis if the server is shutting down
+	if api.srvShutdown.Load() {
+		return
+	}
+
+	// Update the optimistic slot and signal processing of the next slot
 	api.optimisticSlot.Store(headSlot + 1)
+	err = api.redis.BeginProcessingSlot(context.Background(), headSlot+1)
+	if err != nil {
+		api.log.WithError(err).Error("failed to lock redis optimistic processing slot")
+		api.optimisticSlot.Store(0)
```

Comment:
why is the optimistic slot 0 if redis fails to process the slot instead of the previous head?

Reply:
The only place api.optimisticSlot is used is in deciding whether or not to process a submission optimistically:
https://github.com/flashbots/mev-boost-relay/blob/d668875c71f546aa391207b17faf0cb48428699b/services/api/service.go#L2128

We only reach this `BeginProcessingSlot` call once we've received a head event for the previous slot and all optimistic blocks for that slot have have finished processing:
https://github.com/flashbots/mev-boost-relay/blob/d668875c71f546aa391207b17faf0cb48428699b/services/api/service.go#L477

So the only valid block submissions we receive at this point are those with `submission.BidTrace.Slot == headSlot+1`. Whether we run `api.optimisticSlot.Store(0)` or `api.optimisticSlot.Store(prevHeadSlot)`, neither will match the `submission.BidTrace.Slot` of new incoming blocks. Either way, the end result is that if the redis connection fails, the process will not allow optimistic submissions for the upcoming slot.

So while there are a few options for code flow, the current design sets to `headSlot + 1` by default, and resets to `0` as a general "disable all optimistic processing" in case of error.